### ISSUE TITLE
Screenshot file name too long warning when Bytes exceeded

### ISSUE THREAD

**Issue by rafaelcaviquioli** [cypress-io/cypress#2403]
### Current behavior:
Currently I use BDD to document acceptance test cases, each test scenario is well detailed, Cypress uses the description of "it" to create the screenshot file name, but when "it" exceeds 255 characters it is not possible to create the screenshot file.

```
  6) As a quality manager, i would like to query the occurence history to find information about them GIVEN that im onthe search sreen, WHEN i fill in the ocurrence date fields, check the only delayed ones box and click on the filter button, THEN i see the filtered results:
     Error: ENAMETOOLONG: name too long, open '/home/node/cypress/screenshots/tools/us6.spec.js/As a quality manager iwould like to query the occurence history to find information about them -- GIVEN that im on the search sreen WHEN i fill in the ocurrence date fields check the only delayed ones box and click on the filter button THEN i see the filteredresults (failed).png'
```

### Desired behavior:
1. I hope Cypress allows me to change the name of the destination file from the screenshot before it writes it.
1. Or, I hope Cypress will worry about the size of the file before it writes it and limits it to 255 characters.

### Steps to reproduce:
1. Create a test case with "it" description too long, bigger then 255 caracteres.
1. Make a test case broken.
1. Run tests.
1. Look the verbose log, the screenshot not was created.

### Versions
Cypress:    3.0.3
Browser:    Chrome 67

┆Issue is synchronized with this [Jira Bug](https://cypress-io.atlassian.net/browse/TR-90) by [Unito](https://www.unito.io/learn-more)

**Comment by chrisbreiding** [comment]
Thanks for reporting this. We definitely need to catch that error and truncate the file name as needed. I'm hesitant to always truncate at an arbitrary character length since that's going to be OS-dependent and there may be an operating system with an even lower limit.

Speaking of which, on what operating system are you encountering this?

Unfortunately, there's no way to change the screenshot path before it's written to disk. We'd need to add a new plugin event that enables that.

**Comment by rafaelcaviquioli** [comment]
Thank's for response @chrisbreiding , we are running on Ubuntu 16, Ubuntu 18 and Docker image cypress/browsers:chrome67

**Comment by lslv1243** [comment]
+1

**Comment by JulianoGTZ** [comment]
+ 1

**Comment by jennifer-shehane** [comment]
I also got this issue on our internal tests. 🙈 

```
Error: ENAMETOOLONG: name too long, open '/root/cypress-dashboard/cypress/screenshots/project_spec.coffee/Project -- route to project -- permission to view private project -- grace period banner -- billing feature flag ON -- as owner or admin -- not over test recordings -- displays banner with link to billing (failed).png'
```

**Comment by AlDemion** [comment]
I got the same situation with 3.2.0 version

`12:03:21  Wed, 08 May 2019 09:03:21 GMT cypress:server:socket automation:request take:screenshot { titles: [ 'WMED: [STORY] Тестовые сценарии для CI', 'Сценарии:', 'Сценарий 2: Создание обращения, создание медзаписи, привязка обращения к медзаписи', '- Сценарий 2' ], testId: 'r7', simple: true, testFailure: true, capture: 'runner', clip: { x: 0, y: 0, width: 1050, height: 932 }, viewport: { width: 1050, height: 932 }, scaled: true, blackout: [], startTime: '2019-05-08T09:03:20.993Z' }
12:03:21  Wed, 08 May 2019 09:03:21 GMT cypress:server:screenshot (s1) capturing screenshot { titles: [ 'WMED: [STORY] Тестовые сценарии для CI', 'Сценарии:', 'Сценарий 2: Создание обращения, создание медзаписи, привязка обращения к медзаписи', '- Сценарий 2' ], testId: 'r7', simple: true, testFailure: true, capture: 'runner', clip: { x: 0, y: 0, width: 1050, height: 932 }, viewport: { width: 1050, height: 932 }, scaled: true, blackout: [], startTime: '2019-05-08T09:03:20.993Z', specName: 'WMED_UAT_Scenarios_For_CI_spec.js' }
12:03:21  Wed, 08 May 2019 09:03:21 GMT cypress:server:timers clearing timer id 87 from queue { '87': { args: [], ms: 85000, cb: [Function] }, '100': { args: [], ms: 85000, cb: [Function] } }
12:03:21  Wed, 08 May 2019 09:03:21 GMT cypress:server:timers queuing timer id 101 after 85000 ms
12:03:21  Wed, 08 May 2019 09:03:21 GMT cypress:server:timers child received timer id 101
12:03:21  Wed, 08 May 2019 09:03:21 GMT cypress:server:screenshot (s1) save /jenkins-slave/workspace/test-wmed/qa/cypress/wmed_ci/cypress/screenshots/WMED_UAT_Scenarios_For_CI_spec.js/WMED [STORY] Тестовые сценарии для CI -- Сценарии -- Сценарий 2 Создание обращения, создание медзаписи, привязка обращения к медзаписи -- - Сценарий 2 (failed).png
12:03:21  Wed, 08 May 2019 09:03:21 GMT cypress:server:screenshot (s1) clearing 0 cached multipart images
12:03:21  Wed, 08 May 2019 09:03:21 GMT cypress:server:timers clearing timer id 100 from queue { '100': { args: [], ms: 85000, cb: [Function] }, '101': { args: [], ms: 85000, cb: [Function] } }
12:03:21  Wed, 08 May 2019 09:03:21 GMT cypress:server:timers queuing timer id 102 after 85000 ms
12:03:21  Wed, 08 May 2019 09:03:21 GMT cypress:server:timers child received timer id 102
12:03:21  Wed, 08 May 2019 09:03:21 GMT cypress:server:project onMocha fail
12:03:21  Wed, 08 May 2019 09:03:21 GMT cypress:server:reporter got mocha event 'fail' with args: [ { id: 'r7', title: '- Сценарий 2', err: { message: 'ENAMETOOLONG: name too long, open \'/jenkins-slave/workspace/test-wmed/qa/cypress/wmed_ci/cypress/screenshots/WMED_UAT_Scenarios_For_CI_spec.js/WMED [STORY] Тестовые сценарии для CI -- Сценарии -- Сценарий 2 Создание обращения, создание медзаписи, привязка обращения к медзаписи -- - Сценарий 2 (failed).png\'', name: 'Error', stack: 'Error: ENAMETOOLONG: name too long, open \'/jenkins-slave/workspace/test-wmed/qa/cypress/wmed_ci/cypress/screenshots/WMED_UAT_Scenarios_For_CI_spec.js/WMED [STORY] Тестовые сценарии для CI -- Сценарии -- Сценарий 2 Создание обращения, создание медзаписи, привязка обращения к медзаписи -- - Сценарий 2 (failed).png\'\n\n' }, state: 'failed', body: 'function (`

**Comment by jennifer-shehane** [comment]
@AlDemion Thanks so much for providing the logs.

I tested nesting some long suite titles, which passed for me.

I then tried pasting in the exact titles of your test suites - this also passed for me. 

I changed the filename, so that the path would be really long - this also passed for me! 

So, I haven't been able to recreate completely with my guesswork. 

Can you post the full `body` of the error message in the DEBUG log? When you pasted, you cut it off at `state: 'failed', body: function (` - I need the info printed here to see the stack of where this is being thrown.

**Comment by kuznetsovaOk** [comment]
Hello! I've come across with the issue that even after Cypress truncate extra letters from file name I still get ENAMETOOLONG  error in CI. I figured out that the cause of that is using Cyrillic letters in Cypress tests. My filesystem has file name limit in 255 bytes (not characters). One Cyrillic letter takes 2 bytes instead of 1 byte. When Cypress checks if the file name is not too long it seems to be checking characters but not bytes. So it happens that file name with Cyrillic letter pass the Cypress validation but not file system validation. 
@AlDemion maybe you have the same issue. I've checked how many characters vs bytes in your example in apfs:
`echo "WMED [STORY] Тестовые сценарии для CI -- Сценарии -- Сценарий 2 Создание обращения, создание медзаписи, привязка обращения к медзаписи -- - Сценарий 2 (failed).png" | wc -m
     164
`
`echo "WMED [STORY] Тестовые сценарии для CI -- Сценарии -- Сценарий 2 Создание обращения, создание медзаписи, привязка обращения к медзаписи -- - Сценарий 2 (failed).png" | wc -c
     268
`
And it seems that your filename might actually exceed your filesystem limits

**Comment by jennifer-shehane** [comment]
@kuznetsovaOk Thanks for the explanation @kuznetsovaOk, this is amazing, thank you!

The max filename length is dependent on the file system that Cypress would be running in and all of them are listed here: https://en.wikipedia.org/wiki/Comparison_of_file_systems#Limits

But yeah, I think the key takeaway here is that chars !== bytes. Reopening to address bytes issues.

**Comment by adrianpaterson** [comment]
When is this issue expected to be resolved? This is a breaking issue for build pipelines so should be seen as a high priority. Not sure why in over a year it hasn't yet been fixed!

**Comment by AlDemion** [comment]
Yeah @kuznetsovaOk You got the point. It seems that all the problems in the file system limits. But I can't change the filesystem, so one hope is that the issue will be fixed by truncating the filename or by using a randomly generated string for the screenshot name.

**Comment by flotwig** [comment]
The patch should be simple enough, if someone having this issue wants to open a PR. We most likely just need to use `Buffer.byteLength` instead of `String.prototype.length` here:

https://github.com/cypress-io/cypress/blob/98063aec64fe42df6680f227a2aaf5692b73af88/packages/server/lib/screenshots.coffee#L329

**Comment by jbergknoff-rival** [comment]
We're running into this in the `cypress/included:4.2.0` container with no Cyrillic characters. The breakdown from the `ENAMETOOLONG` error message is like this:

```
<96 character path to spec file>/<131 character spec title> -- before all hook (failed).png
```

A total of 260 characters. Is there any workaround?

**Comment by flotwig** [comment]
The original issue opened by @rafaelcaviquioli was fixed, by this PR: #2635

The issue @AlDemion and @kuznetsovaOk are running in to seems to be caused by the difference between `String.length` and `Buffer.byteLength` as described here: https://github.com/cypress-io/cypress/issues/2403#issuecomment-551872521

I've opened a PR that truncates filenames to a maximum of *bytes* on all operating systems.

---

@jbergknoff-rival you can try building from my PR and see if it fixes your issue, or if you want, you can share the filenames that are causing problems and I can try to see. Any Unicode character could cause this problem.

If you don't have any Unicode in your filenames, could you also share [debug logs][debug] from the time that issue occurs?

[debug]: https://docs.cypress.io/guides/guides/debugging.html#Print-DEBUG-logs
